### PR TITLE
[libc][math] Implement fast pass for double precision pow function with up to 1ULP error.

### PR DESCRIPTION
**Pull request by lntue**

### CODE REVIEW

File: libc/test/src/math/pow_test.cpp [hunk 1]
```diff
@@ -0,0 +1,114 @@
+//===-- Unittests for pow ------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "hdr/math_macros.h"
+#include "src/__support/FPUtil/FPBits.h"
```
Comment (overmighty) [inline libc/test/src/math/pow_test.cpp:10]: I don't see hdr/math_macros.h being used, and the `FPBits` type being used is the `using FPBits = LIBC_NAMESPACE::fputil::FPBits<T>;` from `FPTest`.

```suggestion
```

File: libc/test/src/math/pow_test.cpp [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+  auto test = [&](mpfr::RoundingMode rounding_mode) {
+    mpfr::ForceRoundingMode __r(rounding_mode);
+    if (!__r.success)
+      return;
+
+    uint64_t fails = 0;
+    uint64_t count = 0;
+    uint64_t cc = 0;
+    double mx, my, mr = 0.0;
```
Comment (overmighty) [inline libc/test/src/math/pow_test.cpp:54]: This only initializes `mr` to 0.0.

**Review by jhuber6 — APPROVED** [review]
Passes the GPU tests.

File: libc/test/src/math/smoke/pow_test.cpp [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "hdr/math_macros.h"
+#include "src/__support/FPUtil/FPBits.h"
+#include "src/math/pow.h"
+#include "test/UnitTest/FPMatcher.h"
+#include "test/UnitTest/Test.h"
```
Comment (overmighty) [inline libc/test/src/math/smoke/pow_test.cpp:13]:
```suggestion
#include "hdr/fenv_macros.h"
#include "src/math/pow.h"
#include "test/UnitTest/FPMatcher.h"
#include "test/UnitTest/Test.h"
```

File: libc/test/src/math/smoke/CMakeLists.txt [hunk 4]
```diff
@@ -3695,6 +3695,16 @@ add_fp_unittest(
...
 
+add_fp_unittest(
+  pow_test
+  SUITE
+    libc-math-smoke-tests
+  SRCS
+    pow_test.cpp
+  DEPENDS
+    libc.src.math.pow
+)
```
Comment (overmighty) [inline libc/test/src/math/smoke/CMakeLists.txt:3706]:
```suggestion
  DEPENDS
    libc.hdr.fenv_macros
    libc.src.math.pow
)
```

File: libc/test/src/math/smoke/pow_test.cpp [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+using LIBC_NAMESPACE::fputil::testing::ForceRoundingMode;
+using LIBC_NAMESPACE::fputil::testing::RoundingMode;
+
+TEST_F(LlvmLibcPowTest, SpecialNumbers) {
+  constexpr double neg_odd_integer = -3.0;
+  constexpr double neg_even_integer = -6.0;
+  constexpr double neg_non_integer = -1.1;
+  constexpr double pos_odd_integer = 5.0;
+  constexpr double pos_even_integer = 8.0;
+  constexpr double pos_non_integer = 1.1;
```
Comment (overmighty) [inline libc/test/src/math/smoke/pow_test.cpp:25]: Nit: `constexpr` variable names should be in uppercase.

File: libc/src/math/generic/pow.cpp [hunk 6]
```diff
@@ -0,0 +1,490 @@
...
+#include "src/__support/FPUtil/FPBits.h"
+#include "src/__support/FPUtil/PolyEval.h"
+#include "src/__support/FPUtil/double_double.h"
+#include "src/__support/FPUtil/multiply_add.h"
+#include "src/__support/FPUtil/nearest_integer.h"
+#include "src/__support/FPUtil/rounding_mode.h"
+#include "src/__support/FPUtil/sqrt.h" // Speedup for pow(x, 1/2) = sqrt(x)
+#include "src/__support/common.h"
+#include "src/__support/macros/config.h"
+#include "src/__support/macros/optimization.h" // LIBC_UNLIKELY
```
[the remark this answers is not in the document]
Comment (lntue) [reply libc/src/math/generic/pow.cpp:20]: Done.

File: libc/src/math/generic/CMakeLists.txt [hunk 7]
```diff
@@ -1552,6 +1552,26 @@ add_entrypoint_object(
...
+  DEPENDS
+    .common_constants
+    libc.src.__support.FPUtil.fenv_impl
+    libc.src.__support.FPUtil.fp_bits
+    libc.src.__support.FPUtil.multiply_add
+    libc.src.__support.FPUtil.nearest_integer
+    libc.src.__support.FPUtil.polyeval
+    libc.src.__support.FPUtil.rounding_mode
+    libc.src.__support.FPUtil.sqrt
+    libc.src.__support.macros.optimization
```
[the remark this answers is not in the document]
Comment (lntue) [reply libc/src/math/generic/CMakeLists.txt:1570]: Done.

File: libc/src/math/generic/pow.cpp [hunk 8]
```diff
@@ -0,0 +1,490 @@
...
+    }
+
+    if (y_a > 0x43d7'4910'd52d'3052) {
+      if (y_a >= 0x7ff0'0000'0000'0000) {
+        // y is inf or nan
+        if (y_mant != 0)
+          // y is NaN
+          // pow(1, NaN) = 1
+          // pow(x, NaN) = NaN
+          return (x_u == FPBits::one().uintval()) ? 1.0 : y;
```
[the remark this answers is not in the document]
Comment (lntue) [reply libc/src/math/generic/pow.cpp:247]: Done.

File: libc/src/math/generic/pow.cpp [hunk 9]
```diff
@@ -0,0 +1,490 @@
...
+
+        // Now y is +-Inf
+        if (x_abs.is_nan())
+          // pow(NaN, +-Inf) = NaN
+          return x;
+
+        if (x_a == 0x3ff0'0000'0000'0000)
+          // pow(+-1, +-Inf) = 1.0
+          return 1.0;
```
[the remark this answers is not in the document]
Comment (lntue) [reply libc/src/math/generic/pow.cpp:256]: Done.

File: libc/src/math/generic/pow.cpp [hunk 10]
```diff
@@ -0,0 +1,490 @@
...
+      // large enough exponent but not too large, so that the computations
+      // won't be overflow in double precision.
+      y = y_sign ? -0x1.0p100 : 0x1.0p100;
+    }
+
+    // y is finite and non-zero.
+
+    if (x_u == FPBits::one().uintval())
+      // pow(1, y) = 1
+      return 1.0;
```
[the remark this answers is not in the document]
Comment (lntue) [reply libc/src/math/generic/pow.cpp:282]: Done.

File: libc/src/math/generic/pow.cpp [hunk 11]
```diff
@@ -0,0 +1,490 @@
...
+
+    // TODO: Speed things up with pow(2, y) = exp2(y) and pow(10, y) = exp10(y).
+
+    if (x_a == 0) {
+      const bool out_is_neg = x_sign && is_odd_integer(y);
```
[the remark this answers is not in the document]
Comment (lntue) [reply libc/src/math/generic/pow.cpp:287]: Done.

File: libc/test/src/math/pow_test.cpp [hunk 12]
```diff
@@ -0,0 +1,114 @@
+//===-- Unittests for pow ------------------------------------------------===//
```
[the remark this answers is not in the document]
Comment (lntue) [reply libc/test/src/math/pow_test.cpp:1]: Done.

File: libc/test/src/math/pow_test.cpp (hunk 1 above)
Comment (lntue) [reply libc/test/src/math/pow_test.cpp:10]: Done.

File: libc/test/src/math/pow_test.cpp (hunk 2 above)
Comment (lntue) [reply libc/test/src/math/pow_test.cpp:54]: Done.

File: libc/test/src/math/smoke/CMakeLists.txt (hunk 4 above)
Comment (lntue) [reply libc/test/src/math/smoke/CMakeLists.txt:3706]: Done.

File: libc/test/src/math/smoke/pow_test.cpp (hunk 5 above)
Comment (lntue) [reply libc/test/src/math/smoke/pow_test.cpp:25]: Done.

File: libc/src/math/generic/pow.cpp [hunk 13]
```diff
@@ -0,0 +1,490 @@
...
+  uint64_t sign = 0;
+
+  ///////// BEGIN - Check exceptional cases ////////////////////////////////////
+
+  // The double precision number that is closest to 1 is (1 - 2^-53), which has
+  //   log2(1 - 2^-53) ~ -1.715...p-53.
+  // So if |y| > 1075 / log2(1 - 2^-53), and x is finite:
+  //   |y * log2(x)| = 0 or > 1075.
+  // Hence x^y will either overflow or underflow if x is not zero.
+  if (LIBC_UNLIKELY(y_mant == 0 || y_a > 0x43d74910d52d3052 ||
```
[the remark this answers is not in the document]
Comment (lntue) [reply libc/src/math/generic/pow.cpp:226]: `|y| > 1075 / log2(1 - 2^-53)`

File: libc/src/math/generic/pow.cpp [hunk 14]
```diff
@@ -0,0 +1,497 @@
...
+      x_mant = FPBits(x * 0x1.0p64).get_mantissa();
+    }
+
+    // x is finite and negative, and y is a finite integer.
+    if (x_sign) {
+      if (is_integer(y)) {
+        x = -x;
+        if (is_odd_integer(y))
+          // sign = -1.0;
+          sign = 0x8000'0000'0000'0000ULL;
```
Comment (overmighty) [inline libc/src/math/generic/pow.cpp:330]: Nit: some of the hex integer literals in this file use `'` to separate nibbles, some don't, some have uppercase A-F digits/letters, some have lowercase a-f, this literal has the `ULL` suffix, others don't.

**Review by overmighty — APPROVED** [review]
LGTM with a few more nits.

File: libc/src/math/generic/pow.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (overmighty) [reply libc/src/math/generic/pow.cpp:226]: Shouldn't it be `|y| > |1075 / log2(1 - 2^-53)|`?

File: libc/src/math/generic/pow.cpp [hunk 15]
```diff
@@ -0,0 +1,497 @@
...
+        // pow (|x| < 1, +inf) = 0.0f
+        // pow (|x| > 1, -inf) = 0.0f
+        // pow (|x| > 1, +inf) = +inf
+        return ((x_a < FPBits::one().uintval()) == y_sign)
+                   ? FPBits::inf().get_val()
+                   : 0.0;
+      }
+      // x^y will be overflow / underflow in double precision.  Set y to a
+      // large enough exponent but not too large, so that the computations
+      // won't be overflow in double precision.
```
[the remark this answers is not in the document]
Comment (lntue) [reply libc/src/math/generic/pow.cpp:280]: Done.

File: libc/src/math/generic/pow.cpp (hunk 14 above)
Comment (lntue) [reply libc/src/math/generic/pow.cpp:330]: Done.

File: libc/src/math/generic/pow.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (lntue) [reply libc/src/math/generic/pow.cpp]: you are right!  I was careless with the signs.

**Review by overmighty — APPROVED** [review]

**Pull request merged**